text/0265-event-vision.md

```diff
@@ -0,0 +1,91 @@
+
+### What are events?
+
+Events are a type of log which has a required event name and a specific structure which is implied by that event name.
+
+The event name and its implied structure make events much more suitable for observability compared to generic logs.
+
+Because of this, the OpenTelemetry project wants to encourage the transition from generic logs to events.
+
+### OTLP
+
+Since events are a type of log, they share the same OTLP data structure and OTLP pipeline.
+
+### API
+
+OpenTelemetry should have an Event API. This will help to promote the distinction between events and generic logs,
+and encourage the use of events over generic logs.
+
+### Interoperability with generic logging libraries
+
+It should be possible to send events from the Event API to a generic logging library (e.g. Log4j).
+This allows users to integrate events from the Event API into an existing (non-OpenTelemetry) log stream.
+
+Note: If a user chooses to send events from the Event API to a generic logging library, and they have
+also chosen to send the logs from their generic logging library to the OpenTelemetry Logging SDK, then they should
+avoid sending events from the Event API directly to the OpenTelemetry Logging SDK since that would lead to duplicate
+capture of events that were sent from the Event API.
+
+It should be possible to bypass the Event API entirely and emit Events directly via an existing language-specific logging libraries.
+This helps reinforce the idea that events are just a specific type of log.
+
+Note: Because of this, generic event processors should be implemented as Log SDK processors.
+
+OpenTelemetry will recommend that
+[instrumentation libraries](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/glossary.md#instrumentation-library)
+use the Event API over emitting events via a generic logging library in order to give users a simple and consistent
+onboarding story that doesn't involve mixing the two approaches.
+
+OpenTelemetry will recommend that application developers also use the Event API over emitting events via a generic
+logging library since it avoid accidentally emitting logs which lack an event name or are unstructured.
+
+Also, recommending the Event API over emitting events via a generic logging library makes for a clearer overall
+OpenTelemetry API story - with first class user facing APIs for traces, metrics, and events,
+all suitable for using directly in native instrumentation.
+
+### Relationship to Span Events
+
+Events are intended to replace Span Events in the long-term.
+Span Events will be deprecated to signal that users should prefer Events.
+
+Interoperability between Events and Span Events will be defined in the short-term.
+
+### SDK
+
+The Event SDK needs to support two destinations for events:
+
+* OpenTelemetry Logging SDK
+* Language-specific logging libraries
+
+## Trade-offs and mitigations
+
+TODO
+
+## Prior art and alternatives
+
+TODO
```

Comment:
Can please describe why instead of having "Event API" we cannot simply use "Logs (Bridge) API" directly? The logs/events would simple have attributes (e.g. `event.name`) and instrumentation scope as defined in semantic conventions.

Reply:
> Can you please describe why instead of having "Event API" we cannot simply use "Logs (Bridge) API" directly?

That's an interesting idea. The components in the log bridge API (LoggerProvider, Logger, etc) were at least partially named in a way that would make sense in a future world where the API had a scope expanded to be user facing (i.e. no longer just a bridge).

Today `Logger` has a single operation: [emit a LogRecord](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#emit-a-logrecord)

You could imagine adding an additional operation for `emit an Event` and adjusting the phrasing so that `emit a LogRecord` is non-user user facing, reserved for bridging, and `emit an Event` is user facing.

Reply:
I could also imagine a method on `LogRecord` like `asEvent(string name)` that would be just adding an `event.name` string attribute.

However, for sake of instrumentation libraries, I do not think we even need that. What is a problem just adding the attribute explicitly like all other attributes?

I was thinking that it could nice for "business code" so that they can easily emit events. However, even then just making a "facade" for adding event name would not give much if the user is not familiar with the concepts of semantic conventions and schemes. For people understanding these concepts adding an `event.name` attribute  should not be an issue. 

Also, I do not think we should "restrict" semantic logging to only events.

Reply:
I think the moment we document `Logger` as user-facing, we can no longer keep one of its methods non-user facing (but technically public). 

With this proposal, we'd make all logging API public and with @jsuereth analogy https://github.com/open-telemetry/oteps/pull/265#discussion_r1742560611 it's not a bad thing at all.

Reply:
nit: `Logger` is already user-facing. However, it is should be used in log bridges. Excerpt Go docs:

> Package log provides the OpenTelemetry Logs Bridge API.
>
> This package is intended to be a bridge between existing logging libraries and OpenTelemetry. It is not designed to be a logging API itself.

But nothing prevents developers from using it.

Reply:
I also fear the confusion that we would introduce for users when Logs SDK would be used to e.g. process events emitted via Events API.

Reply:
> I also fear the confusion that we would introduce for users when Logs SDK would be used to e.g. process events emitted via Events API.

could you elaborate? events are logrecords with an extra attribute so there should be no surprises.

Reply:
> events are logrecords with an extra attribute so there should be no surprises.

Do you think it would be obvious for end-users who just want to emit events using Event API?

If this would be "no surprise" then it would be totally fine to ask end-users to add the attribute by themselves.

Reply:
So the concern is that if we have Event API, then it's logical to have Event SDK so it's less surprising that events are logs with extra attribute?

I assume having `Logger.emitLog` and `Logger.emitEvent` would resolve it?

Reply:
> I assume having Logger.emitLog and Logger.emitEvent would resolve it?

Yes, if it is in Logs API (and not Events API).

However, I do not find it worth to add a method where the only difference from `emitLog` would be that it also adds an attribute.

Reply:
Moreover, what would we do with events produced via logging libraries which are bridged via Logs API (logs with explicitly added `event.name` attribute/field) ? Aren't they also events?

Reply:
>  libraries which are bridged via Logs API (logs with explicitly added event.name attribute/field) ? Aren't they also events

yes, they are events according to event definition

> If a logging library is capable of creating logs which correctly map
> to the Event data model, logging in this manner is also an acceptable way to
> create Events.

https://github.com/open-telemetry/opentelemetry-specification/blob/625e0f8500f2aea48982a943997c991ff41ee9bb/specification/logs/event-api.md?plain=1#L79

Reply:
If we only need to publish `LogRecord` with different flavors of parameters, like publishing events, having a dedicated provider or API doesn't make much sense to me.

This is the current design:

```mermaid
graph TD
    A[LoggerProvider] -->|Get| B(Logger)
    B -->|Emit| C(LogRecord)

    D[EventLoggerProvider] -->|Get| E(EventLogger)
    E -->|Emit| F(LogRecord)
```

I propose we get rid of `EventLoggerProvider`:

```mermaid
graph TD
    A[LoggerProvider] -->|Get| B(Logger)
    B -->|Emit| C(LogRecord)

    A -->|Get| E(EventLogger)
    E -->|Emit| F(LogRecord)
```

Or, even better, take the `EventLogger` as a wrapper around the `Logger` (I prefer this):

```mermaid
graph TD
    A[LoggerProvider] -->|Get| B(Logger)
    B -->|Emit| C(LogRecord)

    B -. Wrapped .-> E(EventLogger)
    E -->|EmitEvent| C
```

The pseudocode would be:

```go
lp = log.NewLoggerProvider()
otel.SetLoggerProvider(lp)

l = lp.Logger()

el := EventLogger(l)
el.EmitEvent()
```

This looks simple to understand and easy to maintain. Maintainers can focus more on existing log implementation instead of setting up a new set of APIs.

Reply:
💯 👍

Reply:
leaving naming aside, why not

```mermaid
graph TD
    A[LoggerProvider] -->|Get| B(Logger)
    B -->|Emit| C(LogRecord)
    B -->|EmitEvent| C
```

?

Reply:
> leaving naming aside, why not

This is also good, as it might improve performance by reducing the calling chain. However, the `Logger` is a bridge that is not intended to be called by application developers directly. Having `Emit` aside from `EmitEvent`, intended to be called by application developers, feels strange to me.

> The Logs Bridge API is not intended to be called by application developers directly, and SHOULD include documentation that discourages direct use.

https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#artifact-naming

This approach also needs to change [Logs Bridge API](https://opentelemetry.io/docs/specs/otel/logs/bridge-api/), a stable doc, to include `EmitEvent`.

---

The wrapper design can separate `Emit` and `EmitEvent`. So, we do not need to touch the `Logs Bridge API`; leave the `Event API` to only have `EventLogger` to state, and it is a wrapper of `Logger`.

---

Both designs are good to me. I slightly prefer the wrapper one.

Reply:
> Or, even better, take the `EventLogger` as a wrapper around the `Logger` (I prefer this):

Indeed a lot better than adding Event API 👍 However, why not simply?


```go

lp = log.NewLoggerProvider()
otel.SetLoggerProvider(lp)

l = lp.Logger()

el := EventLogger(l)
ev := Event{} // or ev := Record{}? would we need to add Event type which would be almost the same as Record?
el.EmitEvent("special-event", ev)
```

vs

```go
lp = log.NewLoggerProvider()
otel.SetLoggerProvider(lp)

l = lp.Logger()

ev := Record{}
ev.AddAttribute(semconv.EventName("special-event")) r.AsEvent
l.Emit(ev)
```

Alternatively, we can add a method/helper function which makes a log record an event e.g.

```go
lp = log.NewLoggerProvider()
otel.SetLoggerProvider(lp)

l = lp.Logger()

ev := Record{}
ev.AsEvent("special-event")
l.Emit(ev)
```

Reply:
> leaving naming aside, why not

Instrumentation libraries want emit events and not  "raw logs". Thus they will use EventLogger. They do not need `Emit`.
Log bridges want to emit "raw logs" - they do not need `EmitEvent`.
Also, the API implementation would not need to implement `EventLogger` (as this is just a facade). 

Loose reference: https://martinfowler.com/bliki/RoleInterface.html

Reply:
> However, why not simply

Yeah, it seems more straightforward. After all, developers can access the `Logger` if they want.

Then, we can add a helper function to construct an "Event" later based on user feedback.

---

I feel that by **removing this statement**, we can solve all of these Log issues. 🤣

> The Logs Bridge API is not intended to be called by application developers directly, and SHOULD include documentation that discourages direct use.

https://opentelemetry.io/docs/specs/otel/logs/bridge-api/#artifact-naming

Reply:
> I feel that by removing this statement, we can solve all of these Log issues. 🤣

The log bridge API elements were named to allow for evolution into a user facing log API if needed. So this definitely seems feasible and an allowable change. From that [link](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md#artifact-naming):

> in the event OpenTelemetry were to add a user facing API, the Logs Bridge API would be a natural starting point. Therefore, Log Bridge API artifact, package, and class names MUST NOT include the terms "bridge", "appender", or any other qualifier that would prevent evolution into a user facing API.

Reply:
The Event API and the LogBridge API should stay separate, as they represent different parts of our logging system.

Otel has two goals when it comes to logging:

* Consume all of the end users' existing logs.
* Produce our own highly opinionated logs, which will replace span events.

These goals are at the opposite ends of the spectrum. When consuming existing logs we want to be as flexible as possible, when creating our own logs we want to be as rigorous as possible. That's why we have both the LogBridge API and the Event API.

The LogBridge is intended to be used as a log sink for exporting logs of any kind. This includes the events – by default, the Event SDK uses the LogBridge as their event exporter. Optionally, users may want to use SpanEvents as their event exporter. Or they may not want to do any processing at all, and just send the events directly to their application logging system.

Keeping the EventProvider and LoggerProvider separate from each other avoids most of the common pitfalls relating to circular log pipelines and dependencies. For example, users may want to configure their application logger to export data to Otel via the LogBridge. At the same time, they may want all Otel instrumentation (which uses events) to be configured to send data directly to their application logger.  So events flow into the application logger, then all logs flow into the log bridge. Alternatively, events might need a different logging path from regular application logs, in order to preserve their structure and avoid unnecessary processing. Having separate configurations and processor pipelines for the Event API and LogBridge API makes it possible for end users to stay in control of their processing.

Reply:
> when creating our own logs we want to be as rigorous as possible

is it not the purpose of semantic conversations?

> Having separate configurations and processor pipelines for the Event API and LogBridge API makes it possible for end users to stay in control of their processing.

Then you suggest that these should have separate SDKs? Wouldn't this contradict with the idea that events may be emitted by using logging libraries as noted in https://github.com/open-telemetry/oteps/pull/265#discussion_r1745891158. Similar things could be achieved by creating a routing log record processor without having the need to create new API and SDK.

Reply:
Promoting the existing logging APIs causes a few problems, long-term, I think.

First, and most problematic imo, is that it dramatically increases the 'surface area' of the SDK from a maintainability perspective without actually giving us much in return. The point of the bridge is that if you have legacy logging APIs, you can get more out of those logs by feeding them into the SDK to be associated with context or get normalized attributes. If we walk that back and say 'well, ok, do whatever' then we'd need to add more features to be competitive with existing logging APIs and SDKs.

The second is that we would be promoting logs as a signal that we think people should use. This is a bigger mistake, imo. Traditional logging patterns are, frankly, not terribly useful in the OpenTelemetry model. Events are. API separation makes that crystal clear.

Reply:
trying to summarize here and bring these threads together into a decision: https://github.com/open-telemetry/oteps/pull/265#issuecomment-2334806152

Reply:
Shouldn't  we put a brief description about the alternative of having a separate API and SDK any why it is rejected?

Reply:
done: https://github.com/open-telemetry/oteps/pull/265/commits/fc0381e8d9710cb7afccff2b7eb81fe921aacee3